### ISSUE TITLE
On-chain schemas redundant in content directory?

### ISSUE THREAD

**Issue by bedeho** [Joystream/joystream#58]
# Background

The way the current content directory encoding is done is described here

https://github.com/Joystream/joystream/blob/master/reports/archive/1.md#the-json-approach

In particular, it references that all schema will be stored on chain with unique ids, and that content items will have a reference to one such schema via this id, and then have its actual json encoding stored as well.

# Problem

There appears to be no good reason to store the schemas on-chain. Any production user experience using the content directory cannot be schema-driven, it has to render an experience which is tightly coupled to the schemas it supports at the implementation level. Hence the validation rules carried in the scheme must be in the application to begin with. This is indeed what we are doing now in Pioneer also obviously. Any new schemas introduced cannot be handled by such an application, and it must be updated. Hence there is no good reason to store the schemas on-chain.

Even the schema ids may be redundant for this same reason, as they then are not validated on-chain in any way, which was the main principle for putting things inside the json field in the json proposal.

**Comment by jfinkhaeuser** [comment]
I disagree strongly that there is no good reason, although I do agree that the on-chain schemata aren't utilized fully yet.

1. Production user experiences often are schema-driven, albeit partially. Knowing what schema your data has can give you better hints as to what to render and not to render. Typically, applications distinguish between input schemata they support, and a few rendering schemata. They then convert from input schemata to rendering schemata, and contain only specialized code for rendering the latter. This is how user interfaces stay backwards compatible with old data sets - which is how we intend to handle schema changes in future.

Now schemata don't have to be stored on chain for this to work, but some unique identifier per schema has to be allocated. Storing the schema as well adds accountability - if data claims to follow a schema and doesn't, it can be removed as malicious, permitting simpler user interface implementations.
1. The other purpose of the schema, in the ideal case, is to validate before committing to chain. There is no code for this yet, but it would make for a useful future addition.

To summarize:
- use the chain's registry of valid schemata as a verified source of truth for supported data.
- ideally validate data before committing to chain (runtime extension)

**Comment by bedeho** [comment]
1. I not sure what you mean by rendering schemata here, so hard to follow. 

Perhaps you can give some link which shows some example of this?

2. I guess this is something to keep in mind, but my impression was that we were explicitly not going to do this due to some feedback he had from Parity. 

Given the frequency of this even (every content item upload), and that any member should be able to do it, without staking, this seems like it could get very expensive. You basically have to parse JSON, then apply validation against some sort of parsed representation of the schema. The value of the entire exercise is also unclear, as the entire schema model does not have constraints on the relationship between items, so, for example, a podcast can be claimed as an episode in episodic video content. Both items may indeed respect their claimed schemas. Manual curation will be needed to detect and repair this.

**Comment by jfinkhaeuser** [comment]
1. Any application that presents a user interface based on some input data renders the data according to its schema. I may have been unclear on this.

1. Yes, @siman had feedback from Parity that this isn't a good idea. But we already do a lot of things that Parity doesn't seem to like, like storing text on chain. We either accept that we "break the rules" or we don't, IMHO. 
Parsing JSON and validating against a schema isn't a particularly expensive exercise. That's something that happens on most web requests; it takes a few milliseconds. And there's - theoretically, one would have to find an appropriate library - not much code to implement it, as we'd be e.g. using https://json-schema.org/ .
The value is equivalent to the value of using an RDBMs versus a document store, with the added flexibility that the definition of storage items (i.e. both schemata and the data validated by them) can be modified without runtime upgrades.
